src/main/java/de/tum/cit/aet/artemis/communication/service/ConversationMessagingService.java

```diff
@@ -431,4 +452,78 @@ private Conversation mayUpdateOrDeleteMessageElseThrow(Post existingMessagePost,
     public String getEntityName() {
         return METIS_POST_ENTITY_NAME;
     }
+
+    /**
+     * Configures the search specifications based on the provided filter criteria.
+     *
+     * @param specification     The existing specification to be configured.
+     * @param postContextFilter Filtering and sorting properties for post objects.
+     * @param userId            The id of the user for which the messages should be returned.
+     * @return A Specification object configured with search criteria.
+     */
+    private Specification<Post> configureSearchSpecification(Specification<Post> specification, PostContextFilterDTO postContextFilter, long userId) {
+        return specification
+        // @formatter:off
+            .and(getSearchTextAndAuthorSpecification(postContextFilter.searchText(), postContextFilter.authorIds()))
+            .and(getCourseWideChannelsSpecification(Boolean.TRUE.equals(postContextFilter.filterToCourseWide()), postContextFilter.courseId()))
+            .and(getAnsweredOrReactedSpecification(Boolean.TRUE.equals(postContextFilter.filterToAnsweredOrReacted()), userId))
+            .and(getUnresolvedSpecification(Boolean.TRUE.equals(postContextFilter.filterToUnresolved())))
+            .and(getPinnedSpecification(Boolean.TRUE.equals(postContextFilter.pinnedOnly())))
+            .and(getSortSpecification(true, postContextFilter.postSortCriterion(), postContextFilter.sortingOrder()));
+        // @formatter:on
+    }
+
+    /**
+     * Generates SQL Query via specifications to find and sort Messages
+     *
+     * @param postContextFilter filtering and sorting properties for post objects
+     * @param pageable          paging object which contains the page number and number of records to fetch
+     * @param userId            the id of the user for which the messages should be returned
+     * @return returns a Page of Messages
+     */
+    public Page<Post> findMessages(PostContextFilterDTO postContextFilter, Pageable pageable, long userId) {
+        var specification = getConversationsSpecification(postContextFilter.conversationIds());
+        specification = configureSearchSpecification(specification, postContextFilter, userId);
+        // Fetch all necessary attributes to avoid lazy loading (even though relations are defined as EAGER in the domain class, specification queries do not respect this)
+        return findPostsWithSpecification(pageable, specification);
+    }
+
+    private PageImpl<Post> findPostsWithSpecification(Pageable pageable, Specification<Post> specification) {
```

Comment:
Why are these methods moved to a Service? What califies them for service methods instead of repository methods? 

These are simple data retrieval methods, not performing any business logic. For my understanding these would belong to a repository instead of a service.